UIforETW/Utility.cpp

```diff
@@ -127,71 +276,125 @@ std::wstring ConvertToCRLF(const std::wstring& input)
-	if (result == ERROR_SUCCESS)
-	{
-		RegSetValueEx(key, valueName.c_str(), 0, REG_DWORD, reinterpret_cast<const BYTE*>(&value), sizeof(value));
-		RegCloseKey(key);
-	}
+	ATLVERIFY(::RegOpenKeyExW(root, subkey.c_str(), 0, KEY_ALL_ACCESS, &key) == ERROR_SUCCESS);
+	ATLVERIFY(ERROR_SUCCESS == ::RegSetValueExW(key, valueName.c_str(), 0, REG_DWORD, reinterpret_cast<const BYTE*>(&value), sizeof(value)));
+	ATLVERIFY(::RegCloseKey(key) == ERROR_SUCCESS);
 }
 
-void CreateRegistryKey(HKEY root, const std::wstring& subkey, const std::wstring& newKey)
+void CreateRegistryKey(const HKEY root, const std::wstring& subkey, const std::wstring& newKey)
 {
 	HKEY key;
-	LONG result = RegOpenKeyEx(root, subkey.c_str(), 0, KEY_ALL_ACCESS, &key);
-	if (result == ERROR_SUCCESS)
+	ATLVERIFY(::RegOpenKeyExW(root, subkey.c_str(), 0, KEY_ALL_ACCESS, &key) == ERROR_SUCCESS);
+
+	HKEY resultKey;
+	//TODO: RegCreateKey is depreciated.
+	const LONG createResult = ::RegCreateKeyW(key, newKey.c_str(), &resultKey);
+	UIETWASSERT(createResult == ERROR_SUCCESS);
+	if (createResult == ERROR_SUCCESS)
 	{
-		HKEY resultKey;
-		result = RegCreateKey(key, newKey.c_str(), &resultKey);
-		if (result == ERROR_SUCCESS)
-		{
-			RegCloseKey(resultKey);
-		}
-		RegCloseKey(key);
+		ATLVERIFY(::RegCloseKey(resultKey) == ERROR_SUCCESS);
 	}
+
+	ATLVERIFY(::RegCloseKey(key) == ERROR_SUCCESS);
 }
 
-std::wstring GetEditControlText(HWND hEdit)
+std::wstring GetEditControlText(const HWND hEdit)
 {
-	std::wstring result;
-	int length = GetWindowTextLength(hEdit);
+	const int length = ::GetWindowTextLengthW(hEdit);
 	std::vector<wchar_t> buffer(length + 1);
-	GetWindowText(hEdit, &buffer[0], static_cast<int>(buffer.size()));
+
+	//GetWindowText https://msdn.microsoft.com/en-us/library/windows/desktop/ms633520.aspx
+	//If [GetWindowTextW] succeeds, the return value is the length,
+	//in characters, of the copied string, not including the
+	//terminating null character.
+
+	//If the window has no title bar or text,
+	//[or] if the title bar is empty,
+	//or if the window or control handle is invalid,
+	//the return value is zero. 
+
+	const int charsWritten = ::GetWindowTextW(hEdit, &buffer[0], static_cast<int>(buffer.size()));
+	if (charsWritten == 0)
+	{
+		if (length > 0)
+			debugLastError();
+		return L"";
+	}
+	UIETWASSERT(charsWritten == length);
+	UIETWASSERT((charsWritten) < (int)buffer.size());
+	UIETWASSERT(buffer[charsWritten] == 0);
 	// Double-verify that the buffer is null-terminated.
 	buffer[buffer.size() - 1] = 0;
 	return &buffer[0];
 }
 
+//MultiByteToWideChar: https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072.aspx
+//
+//Remarks:
+//As mentioned in the caution above,
+//the output buffer can easily be overrun
+//if this function is not first called with cchWideChar set to 0
+//in order to obtain the required size. 
+int RequiredNumberOfWideChars(const std::string& text)
+{
+	static_assert( sizeof(std::string::value_type) == 1 == sizeof(text[0]),

```

Comment:
This seems to be asserting that sizeof(char) is 1, which is true by definition. Harmless, mostly.